bin/remote-prover/src/utils.rs

```diff
@@ -1,52 +1,124 @@
+
+    // Add compression flag (1 byte, 0 = no compression)
+    grpc_message.push(0u8);
+
+    // Add message length (4 bytes, big-endian)
+    let msg_len = response_body.len() as u32;
+    grpc_message.extend_from_slice(&msg_len.to_be_bytes());
+
+    // Add the actual message
+    grpc_message.extend_from_slice(&response_body);
+
+    // Create gRPC response headers WITHOUT grpc-status (that goes in trailers)
+    let mut header = ResponseHeader::build(200, None)?;
+    header.insert_header("content-type", "application/grpc".to_string())?;
+
+    session.set_keepalive(None);
+    session.write_response_header(Box::new(header), false).await?;
+    session.write_response_body(Some(grpc_message.into()), false).await?;
+
+    // Send trailers with gRPC status
+    let mut trailers = HeaderMap::new();
+    let status_value = grpc_status.unwrap_or("0"); // Default to "0" (OK)
+    trailers.insert(
+        "grpc-status",
+        status_value.parse().map_err(|_| Error::new(ErrorType::InternalError))?,
+    );
+    session.write_response_trailers(trailers).await?;
+
+    Ok(true)
+}
+
+/// Write a gRPC error response to a Pingora session
+///
+/// This helper function creates a proper gRPC error response with the specified
+/// status code and error message.
+pub async fn write_grpc_error_to_session(
     session: &mut Session,
+    grpc_status: Code,
+    error_message: &str,
 ) -> pingora_core::Result<bool> {
-    // Set grpc-message header to "Too many requests in the queue"
-    // This is meant to be used by a Tonic interceptor to return a gRPC error
-    let mut header = ResponseHeader::build(503, None)?;
-    header.insert_header("grpc-message", "Too many requests in the queue".to_string())?;
-    header.insert_header("grpc-status", RESOURCE_EXHAUSTED_CODE)?;
+    // Create gRPC response headers (always HTTP 200 for gRPC)
+    let mut header = ResponseHeader::build(200, None)?;
+    header.insert_header("content-type", "application/grpc".to_string())?;
+
     session.set_keepalive(None);
-    session.write_response_header(Box::new(header.clone()), true).await?;
+    session.write_response_header(Box::new(header), false).await?;
 
-    let mut error = Error::new(ErrorType::HTTPStatus(503))
-        .more_context("Too many requests in the queue")
-        .into_in();
-    error.set_cause("Too many requests in the queue");
+    // gRPC errors don't have a body, just headers and trailers
+    session.write_response_body(None, false).await?;
 
-    session.write_response_header(Box::new(header), false).await?;
+    // Send trailers with gRPC status and error message
+    let mut trailers = HeaderMap::new();
+    let status_code = (grpc_status as i32).to_string();
+    trailers.insert(
+        "grpc-status",
+        status_code.parse().map_err(|_| Error::new(ErrorType::InternalError))?,
+    );
+
+    trailers.insert(
+        "grpc-message",
+        error_message.parse().map_err(|_| Error::new(ErrorType::InternalError))?,
+    );
```

Comment:
nit: maybe we can extract this out into a small helper function - something like:
```rust
fn build_grpc_trailers(grpc_status: Code, err_message: Option<&str>) -> HeaderMap {
    ...
}
```
Then we'd be able to use it here and also instead of lines 53 - 58 above.